File: src/lib/encryption.ts
import CryptoJS from "crypto-js";

export enum EncryptionMethods {
  AES = "AES",
  TripleDES = "TripleDES",
  DES = "DES",
  Rabbit = "Rabbit",
  RC4 = "RC4",
  base64 = "base64",
}

export interface EnigmaSchema {
  method: EncryptionMethods;
  key?: () => string;
}

export type Method = Exclude<keyof typeof EncryptionMethods, "base64">;

// Helper functions for common tasks
function hashKey(key: string): string {
  return CryptoJS.SHA256(key).toString(CryptoJS.enc.Hex);
}

// Encryption and Decryption methods
function encryptMethod(method: Method, key: string, plaintext: string): string {
  const keyWA = CryptoJS.enc.Hex.parse(key);

  if (method === EncryptionMethods.RC4) {
    return CryptoJS.RC4.encrypt(plaintext, keyWA).toString();
  }

  // Block ciphers (AES, DES, TripleDES)
  const ivWA = CryptoJS.lib.WordArray.random(
    method === EncryptionMethods.AES ? 16 : 8,
  );

  const encrypted = CryptoJS[method].encrypt(plaintext, keyWA, {
    iv: ivWA,
    mode: CryptoJS.mode.CBC,
    padding: CryptoJS.pad.Pkcs7,
  });

  // Prepend IV to ciphertext (Go standard)
  const combined = ivWA.concat(encrypted.ciphertext);
  return CryptoJS.enc.Base64.stringify(combined);
}

function decryptMethod(
  method: Method,
  key: string,
  cipher_text: string,
): string {
  const keyWA = CryptoJS.enc.Hex.parse(key);

  if (method === EncryptionMethods.RC4) {
    return CryptoJS.RC4.decrypt(cipher_text, keyWA).toString(CryptoJS.enc.Utf8);
  }

  // Block ciphers (AES, DES, TripleDES)
  const combinedWA = CryptoJS.enc.Base64.parse(cipher_text);
  const blockSize = method === EncryptionMethods.AES ? 16 : 8;

  // Extract IV from the start
  const ivWA = CryptoJS.lib.WordArray.create(
    combinedWA.words.slice(0, blockSize / 4),
    blockSize,
  );
  const ciphertextWA = CryptoJS.lib.WordArray.create(
    combinedWA.words.slice(blockSize / 4),
    combinedWA.sigBytes - blockSize,
  );

  return CryptoJS[method].decrypt(
    // @ts-ignore
    { ciphertext: ciphertextWA },
    keyWA,
    {
      iv: ivWA,
      mode: CryptoJS.mode.CBC,
      padding: CryptoJS.pad.Pkcs7,
    },
  ).toString(CryptoJS.enc.Utf8);
}

function applyMethod(
  text: string,
  layer: EnigmaSchema,
  isEncrypt: boolean,
): string {
  if (layer.method === EncryptionMethods.base64) {
    return isEncrypt
      ? CryptoJS.enc.Base64.stringify(CryptoJS.enc.Utf8.parse(text))
      : CryptoJS.enc.Base64.parse(text).toString(CryptoJS.enc.Utf8);
  } else if (layer.key) {
    const key = layer.key();
    return isEncrypt
      ? encryptMethod(layer.method, hashKey(key), text)
      : decryptMethod(layer.method, hashKey(key), text);
  }
  throw new Error(`Key function missing for method: ${layer.method}`);
}

export function encode(enigma_schema: EnigmaSchema[], text: string): string {
  return enigma_schema.reduce(
    (cipher_text, layer) => applyMethod(cipher_text, layer, true),
    text,
  );
}

export function decode(
  enigma_schema: EnigmaSchema[],
  encrypted: string,
): string {
  return enigma_schema.reduceRight(
    (plaintext, layer) => applyMethod(plaintext, layer, false),
    encrypted,
  );
}

File: core/function/encryption.go
package function

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/des"
	"crypto/rand"
	"crypto/rc4"
	"crypto/sha256"
	"encoding/base64"
	"errors"
	"fmt"
	"io"
)

type Encryption struct{}

type EncryptionMethod string

const (
	AES       EncryptionMethod = "AES"
	TripleDES EncryptionMethod = "TripleDES"
	DES       EncryptionMethod = "DES"
	RC4       EncryptionMethod = "RC4"
	Base64    EncryptionMethod = "base64"
)

type EnigmaSchema struct {
	Method EncryptionMethod
	Key    func() string
}

// Helper function to hash the key using SHA256
func hashKey(key string) []byte {
	hash := sha256.Sum256([]byte(key))
	return hash[:]
}

// Helper function to reverse strings
func ReverseStrings(s string) string {
	runes := []rune(s)
	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}
	return string(runes)
}

// PKCS7 Padding
func pkcs7Padding(ciphertext []byte, blockSize int) []byte {
	padding := blockSize - len(ciphertext)%blockSize
	padtext := bytes.Repeat([]byte{byte(padding)}, padding)
	return append(ciphertext, padtext...)
}

// PKCS7 Unpadding
func pkcs7Unpadding(origData []byte) ([]byte, error) {
	length := len(origData)
	if length == 0 {
		return nil, errors.New("empty data")
	}
	unpadding := int(origData[length-1])
	if unpadding > length {
		return nil, errors.New("invalid padding")
	}
	return origData[:(length - unpadding)], nil
}

// Generic CBC Encryption
func encryptCBC(block cipher.Block, plaintext string) (string, error) {
	blockSize := block.BlockSize()
	content := pkcs7Padding([]byte(plaintext), blockSize)

	ciphertext := make([]byte, blockSize+len(content))
	iv := ciphertext[:blockSize]
	if _, err := io.ReadFull(rand.Reader, iv); err != nil {
		return "", err
	}

	mode := cipher.NewCBCEncrypter(block, iv)
	mode.CryptBlocks(ciphertext[blockSize:], content)

	return base64.StdEncoding.EncodeToString(ciphertext), nil
}

// Generic CBC Decryption
func decryptCBC(block cipher.Block, ciphertext string) (string, error) {
	data, err := base64.StdEncoding.DecodeString(ciphertext)
	if err != nil {
		return "", err
	}

	blockSize := block.BlockSize()
	if len(data) < blockSize {
		return "", errors.New("ciphertext too short")
	}

	iv := data[:blockSize]
	data = data[blockSize:]

	if len(data)%blockSize != 0 {
		return "", errors.New("ciphertext is not a multiple of the block size")
	}

	mode := cipher.NewCBCDecrypter(block, iv)
	mode.CryptBlocks(data, data)

	res, err := pkcs7Unpadding(data)
	if err != nil {
		return "", err
	}
	return string(res), nil
}

// AES Encryption and Decryption
func encryptAES(key []byte, plaintext string) (string, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}
	return encryptCBC(block, plaintext)
}

func decryptAES(key []byte, ciphertext string) (string, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}
	return decryptCBC(block, ciphertext)
}

// DES Encryption and Decryption
func encryptDES(key []byte, plaintext string) (string, error) {
	block, err := des.NewCipher(key[:8]) // DES uses 8-byte key
	if err != nil {
		return "", err
	}
	return encryptCBC(block, plaintext)
}

func decryptDES(key []byte, ciphertext string) (string, error) {
	block, err := des.NewCipher(key[:8])
	if err != nil {
		return "", err
	}
	return decryptCBC(block, ciphertext)
}

// TripleDES Encryption and Decryption
func encryptTripleDES(key []byte, plaintext string) (string, error) {
	block, err := des.NewTripleDESCipher(key[:24]) // 3DES uses 24-byte key
	if err != nil {
		return "", err
	}
	return encryptCBC(block, plaintext)
}

func decryptTripleDES(key []byte, ciphertext string) (string, error) {
	block, err := des.NewTripleDESCipher(key[:24])
	if err != nil {
		return "", err
	}
	return decryptCBC(block, ciphertext)
}

// RC4 Encryption and Decryption
func encryptRC4(key []byte, plaintext string) (string, error) {
	cipher, err := rc4.NewCipher(key)
	if err != nil {
		return "", err
	}

	ciphertext := make([]byte, len(plaintext))
	cipher.XORKeyStream(ciphertext, []byte(plaintext))

	return base64.StdEncoding.EncodeToString(ciphertext), nil
}

func decryptRC4(key []byte, ciphertext string) (string, error) {
	data, err := base64.StdEncoding.DecodeString(ciphertext)
	if err != nil {
		return "", err
	}

	cipher, err := rc4.NewCipher(key)
	if err != nil {
		return "", err
	}

	plaintext := make([]byte, len(data))
	cipher.XORKeyStream(plaintext, data)

	return string(plaintext), nil
}

// Apply encryption or decryption method
func applyMethod(text string, layer EnigmaSchema, isEncrypt bool) (string, error) {
	if layer.Method == Base64 {
		if isEncrypt {
			return base64.StdEncoding.EncodeToString([]byte(text)), nil
		}
		data, err := base64.StdEncoding.DecodeString(text)
		if err != nil {
			return "", err
		}
		return string(data), nil
	} else if layer.Key != nil {
		key := hashKey(layer.Key())
		switch layer.Method {
		case AES:
			if isEncrypt {
				return encryptAES(key, text)
			} else {
				return decryptAES(key, text)
			}
		case DES:
			if isEncrypt {
				return encryptDES(key, text)
			} else {
				return decryptDES(key, text)
			}
		case TripleDES:
			if isEncrypt {
				return encryptTripleDES(key, text)
			} else {
				return decryptTripleDES(key, text)
			}
		case RC4:
			if isEncrypt {
				return encryptRC4(key, text)
			} else {
				return decryptRC4(key, text)
			}
		}
	}
	return "", fmt.Errorf("Key function missing for method: %s", layer.Method)
}

// Encode applies encryption layers to the input text
func (ref Encryption) Encode(enigmaSchema []EnigmaSchema, text string) (string, error) {
	var err error
	cipherText := text
	for _, layer := range enigmaSchema {
		cipherText, err = applyMethod(cipherText, layer, true)
		if err != nil {
			return "", err
		}
	}
	return cipherText, nil
}

// Decode applies decryption layers to the input cipher text
func (ref Encryption) Decode(enigmaSchema []EnigmaSchema, encrypted string) (string, error) {
	var err error
	plainText := encrypted
	for i := len(enigmaSchema) - 1; i >= 0; i-- {
		plainText, err = applyMethod(plainText, enigmaSchema[i], false)
		if err != nil {
			return "", err
		}
	}
	return plainText, nil
}
